plugins/out_s3/s3.c

```diff
@@ -1114,11 +1164,33 @@ static int upload_data(struct flb_s3 *ctx, struct s3_file *chunk,
         if (ctx->compression == FLB_AWS_COMPRESS_GZIP) {
             flb_free(payload_buf);
         }
-        m_upload->upload_errors += 1;
         /* re-add chunk to list */
         if (chunk) {
-            s3_store_file_unlock(chunk);
             chunk->failures += 1;
+            if (chunk->failures > ctx->ins->retry_limit) {
+                s3_retry_warn(ctx, (char *) chunk->fsf->meta_buf, m_upload->input_name,
+                              chunk->create_time, FLB_FALSE);
+                s3_store_file_delete(ctx, chunk);
+                /*
+                 * part_number initializes with 1, if the number still is 1 which means
+                 * no data is uploaded and this upload file will be deleted , else set
+                 * as complete.
```

Comment:
nit: to be more clear I'd say "else mark it for completion to save already uploaded data."